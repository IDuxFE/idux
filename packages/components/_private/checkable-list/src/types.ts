/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
