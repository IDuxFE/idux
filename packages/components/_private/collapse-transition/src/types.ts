/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
