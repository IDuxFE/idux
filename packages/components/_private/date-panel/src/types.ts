/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
