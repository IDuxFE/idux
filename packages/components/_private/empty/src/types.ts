/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
