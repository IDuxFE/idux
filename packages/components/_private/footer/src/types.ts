/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { ButtonProps } from '@idux/components/button'
import type { DefineComponent, HTMLAttributes, PropType, VNode } from 'vue'

export interface FooterButtonProps extends ButtonProps {
  key?: VKey
  text?: string | VNode
  onClick?: (evt: Event) => void
}

export const footerProps = {
  cancel: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  cancelButton: Object as PropType<ButtonProps>,
  cancelLoading: Boolean,
  cancelText: String,
  cancelVisible: {
    type: Boolean,
    default: true,
  },
  footer: [Boolean, Array, Object] as PropType<boolean | FooterButtonProps[] | VNode>,
  ok: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  okButton: Object as PropType<ButtonProps>,
  okLoading: Boolean,
  okText: String,
} as const

export type FooterProps = ExtractInnerPropTypes<typeof footerProps>
export type FooterPublicProps = ExtractPublicPropTypes<typeof footerProps>
export type FooterComponent = DefineComponent<Omit<HTMLAttributes, keyof FooterPublicProps> & FooterPublicProps>
export type FooterInstance = InstanceType<DefineComponent<FooterProps>>
