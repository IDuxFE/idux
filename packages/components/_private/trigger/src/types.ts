/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { FormSize } from '@idux/components/form'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

import { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'

export const triggerProps = {
  borderless: Boolean as PropType<boolean>,
  clearable: Boolean as PropType<boolean>,
  clearIcon: String as PropType<string>,
  className: String as PropType<string>,
  disabled: Boolean as PropType<boolean>,
  focused: Boolean as PropType<boolean>,
  readonly: Boolean as PropType<boolean>,
  size: String as PropType<FormSize>,
  suffix: String as PropType<string>,
  onClick: [Array, Function] as PropType<MaybeArray<(evt: Event) => void>>,
  onClear: [Array, Function] as PropType<MaybeArray<(evt: MouseEvent) => void>>,
  onFocus: [Array, Function] as PropType<MaybeArray<(evt: FocusEvent) => void>>,
  onBlur: [Array, Function] as PropType<MaybeArray<(evt: FocusEvent) => void>>,
  onKeyDown: [Array, Function] as PropType<MaybeArray<(evt: KeyboardEvent) => void>>,
}

export type TriggerProps = ExtractInnerPropTypes<typeof triggerProps>
export type TriggerPublicProps = ExtractPublicPropTypes<typeof triggerProps>
export type TriggerComponent = DefineComponent<Omit<HTMLAttributes, keyof TriggerPublicProps> & TriggerPublicProps>
export type TriggerInstance = InstanceType<DefineComponent<TriggerProps>>
