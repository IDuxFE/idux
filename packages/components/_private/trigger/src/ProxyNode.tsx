/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { FunctionalComponent } from 'vue'

import { getFirstValidNode } from '@idux/cdk/utils'

const ProxyNode: FunctionalComponent = (_, { slots }) => {
  return getFirstValidNode(slots.default?.())
}

export default ProxyNode
