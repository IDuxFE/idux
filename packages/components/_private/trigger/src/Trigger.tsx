/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass, onBeforeUnmount, onMounted, ref, watch } from 'vue'

import { useSharedFocusMonitor } from '@idux/cdk/a11y'
import { callEmit } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { IxIcon } from '@idux/components/icon'

import { triggerProps } from './types'

export default defineComponent({
  props: triggerProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-trigger`)

    const isDisabled = computed(() => props.disabled)

    const focusMonitor = useSharedFocusMonitor()
    const triggerRef = ref<HTMLElement>()
    onMounted(() => {
      watch(focusMonitor.monitor(triggerRef.value!, true), evt => {
        const { origin, event } = evt
        if (event) {
          if (origin) {
            callEmit(props.onFocus, event)
          } else {
            callEmit(props.onBlur, event)
          }
        }
      })
    })

    onBeforeUnmount(() => focusMonitor.stopMonitoring(triggerRef.value!))

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [`${props.className}`]: !!props.className,
        [prefixCls]: true,
        [`${prefixCls}-disabled`]: isDisabled.value,
        [`${prefixCls}-borderless`]: props.borderless,
        [`${prefixCls}-readonly`]: props.readonly,
        [`${prefixCls}-focused`]: props.focused,
        [`${prefixCls}-${props.size}`]: props.size,
      })
    })

    const handleClick = (evt: Event) => {
      if (isDisabled.value) {
        return
      }

      callEmit(props.onClick, evt)
    }

    const handleKeyDown = (evt: KeyboardEvent) => {
      if (isDisabled.value) {
        return
      }

      callEmit(props.onKeyDown, evt)
    }

    const handleClear = (evt: MouseEvent) => {
      if (isDisabled.value) {
        return
      }

      evt.stopPropagation()
      callEmit(props.onClear, evt)
    }
    const renderSuffix = () => {
      if (!(slots.suffix || props.suffix)) {
        return null
      }

      return (
        <div class={`${mergedPrefixCls.value}-suffix`}>
          {slots.suffix?.() ?? (props.suffix && <IxIcon name={props.suffix}></IxIcon>)}
        </div>
      )
    }
    const renderClearIcon = () => {
      if (!props.clearable || isDisabled.value || (!props.clearIcon && !slots.clearIcon)) {
        return null
      }

      return (
        <span class={`${mergedPrefixCls.value}-clear-icon`} onClick={handleClear}>
          {slots.clearIcon ? slots.clearIcon() : props.clearIcon && <IxIcon name={props.clearIcon}></IxIcon>}
        </span>
      )
    }

    return () => (
      <div ref={triggerRef} class={classes.value} onClick={handleClick} onKeydown={handleKeyDown}>
        {slots.default?.()}
        {renderSuffix()}
        {renderClearIcon()}
      </div>
    )
  },
})
