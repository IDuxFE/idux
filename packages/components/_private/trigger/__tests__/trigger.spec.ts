import { MountingOptions, mount } from '@vue/test-utils'
import { h } from 'vue'

import { renderWork } from '@tests'

import Trigger from '../src/Trigger'
import { TriggerProps } from '../src/types'

describe('Trigger', () => {
  const TriggerMount = (options?: MountingOptions<Partial<TriggerProps>>) =>
    mount(Trigger, { ...(options as MountingOptions<TriggerProps>) })

  renderWork<TriggerProps>(Trigger, {
    props: {},
  })

  test('borderless work', async () => {
    const wrapper = TriggerMount({ props: { borderless: true } })

    expect(wrapper.classes()).toContain('ix-trigger-borderless')

    await wrapper.setProps({ borderless: false })

    expect(wrapper.classes()).not.toContain('ix-trigger-borderless')
  })

  test('focused work', async () => {
    const wrapper = TriggerMount({ props: { focused: true } })

    expect(wrapper.classes()).toContain('ix-trigger-focused')

    await wrapper.setProps({ focused: false })

    expect(wrapper.classes()).not.toContain('ix-trigger-focused')
  })

  test('readonly work', async () => {
    const wrapper = TriggerMount({ props: { readonly: true } })

    expect(wrapper.classes()).toContain('ix-trigger-readonly')

    await wrapper.setProps({ readonly: false })

    expect(wrapper.classes()).not.toContain('ix-trigger-readonly')
  })

  test('disabled work', async () => {
    const wrapper = TriggerMount({ props: { disabled: true } })

    expect(wrapper.classes()).toContain('ix-trigger-disabled')

    await wrapper.setProps({ disabled: false })

    expect(wrapper.classes()).not.toContain('ix-trigger-disabled')
  })

  test('clearable work', async () => {
    const wrapper = TriggerMount({ props: { clearable: true, clearIcon: 'clear' } })

    expect(wrapper.find('.ix-trigger-clear-icon').exists()).toBeTruthy()

    await wrapper.setProps({ clearable: false })

    expect(wrapper.find('.ix-trigger-clear-icon').exists()).toBeFalsy()
  })

  test('clearIcon work', async () => {
    const wrapper = TriggerMount({ props: { clearable: true, clearIcon: 'clear' } })

    expect(wrapper.find('.ix-icon-clear').exists()).toBeTruthy()

    await wrapper.setProps({ clearIcon: 'close' })

    expect(wrapper.find('.ix-icon-clear').exists()).toBeFalsy()
    expect(wrapper.find('.ix-icon-close').exists()).toBeTruthy()
  })

  test('slot clearIcon work', async () => {
    const wrapper = TriggerMount({
      props: { clearable: true },
      slots: {
        clearIcon: () => h('span', { class: 'custom-clear-icon-slot' }),
      },
    })

    expect(wrapper.find('.custom-clear-icon-slot').exists()).toBeTruthy()
  })

  test('suffix work', async () => {
    const wrapper = TriggerMount({ props: { suffix: 'info' } })

    expect(wrapper.find('.ix-trigger-suffix').exists()).toBeTruthy()
    expect(wrapper.find('.ix-icon-info').exists()).toBeTruthy()
  })

  test('slot suffix work', async () => {
    const wrapper = TriggerMount({
      props: { clearable: true },
      slots: {
        clearIcon: () => h('span', { class: 'custom-suffix-slot' }),
      },
    })

    expect(wrapper.find('.custom-suffix-slot').exists()).toBeTruthy()
  })

  test('onClick work', async () => {
    const onClick = vi.fn()
    const wrapper = TriggerMount({ props: { onClick } })

    await wrapper.trigger('click')
    expect(onClick).toBeCalled()

    onClick.mockClear()

    await wrapper.setProps({ disabled: true })
    await wrapper.trigger('click')
    expect(onClick).not.toBeCalled()
  })

  test('onClear work', async () => {
    const onClear = vi.fn()
    const wrapper = TriggerMount({ props: { clearable: true, clearIcon: 'clear', onClear } })

    await wrapper.find('.ix-trigger-clear-icon').trigger('click')
    expect(onClear).toBeCalled()

    await wrapper.setProps({ disabled: true })
    expect(wrapper.find('.ix-trigger-clear-icon').exists()).toBeFalsy()
  })

  test('onKeyDown work', async () => {
    const onKeyDown = vi.fn()

    const wrapper = TriggerMount({ props: { onKeyDown } })

    await wrapper.trigger('keydown')
    expect(onKeyDown).toBeCalled()

    onKeyDown.mockClear()

    await wrapper.setProps({ disabled: true })
    await wrapper.trigger('keydown')
    expect(onKeyDown).not.toBeCalled()
  })
})
