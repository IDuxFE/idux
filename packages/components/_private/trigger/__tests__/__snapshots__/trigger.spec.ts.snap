// Vitest Snapshot v1

exports[`Trigger > render work 1`] = `
"<div class=\\"ix-trigger\\">
  <!---->
  <!---->
  <!---->
</div>"
`;
